src/dstack/_internal/server/services/storage/base.py

```diff
@@ -0,0 +1,25 @@
+from abc import ABC
+from typing import Optional
+
+
+class BaseStorage(ABC):
+    def upload_code(
+        self,
+        project_id: str,
+        repo_id: str,
+        code_hash: str,
+        blob: bytes,
+    ):
+        raise NotImplementedError
+
+    def get_code(
+        self,
+        project_id: str,
+        repo_id: str,
+        code_hash: str,
+    ) -> Optional[bytes]:
+        raise NotImplementedError
```

Comment:
Required method should be decorated with `@abstractmethod` and have empty body (`pass` or `...`) instead of `raise NotImplementedError`